File: ios/Runner/DownloadTask.swift
import Foundation

enum DownloadStatus: String, Codable {
    case queued
    case active
    case paused
    case completed
    case failed
}

struct DownloadTask: Codable {
    let id: String
    let url: URL
    let filename: String
    var status: DownloadStatus
    var progress: Double
    var speed: Double
    var downloadedBytes: Int64
    var totalBytes: Int64
    var resumeData: Data?
    var error: String?

    init(id: String, url: URL, filename: String) {
        self.id = id
        self.url = url
        self.filename = filename
        self.status = .queued
        self.progress = 0
        self.speed = 0
        self.downloadedBytes = 0
        self.totalBytes = 0
    }
}

File: lib/models/download_task.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'download_task.freezed.dart';
part 'download_task.g.dart';

enum DownloadStatus {
  queued,
  active,
  paused,
  completed,
  failed,
  cancelled
}

@freezed
class DownloadTask with _$DownloadTask {
  const factory DownloadTask({
    required String id,
    required String url,
    required String filename,
    required DownloadStatus status,
    @Default(0.0) double progress,
    @Default(0.0) double speed, // bytes per second
    @Default('') String eta,
    @Default(0) int downloadedBytes,
    @Default(0) int totalBytes,
    String? errorMessage,
  }) = _DownloadTask;

  factory DownloadTask.fromJson(Map<String, dynamic> json) => _$DownloadTaskFromJson(json);
}
